### PR TITLE
Oculars plugin: Introduce a 5-value pluginMode property

### LINKED ISSUE #356: finderscope option in Ocular plugin
_Original report by Stub Mandrel:_ https://bugs.launchpad.net/stellarium/+bug/1511391

To create a finderscope in the ocular plugin you have to reverse engineer the focal length from its magnification and aperture.

it would be better if finderscopes were a separate set of options defined by aperture and magnification, as this would also allow you to 'flip' between the scope view and the finder view with one click.

Final thought, if the finder and scope views could be shown side by side it would be great.

### LINKED ISSUE #1509: Telrad view ignore magnitude limit
Hello,
I don't know if it's a bug but it's very special.
Viewing with the telrad mode disables the magnitude limit, and worse when exiting this mode, the limit is not re-activated.
I am using the latest version on windows 10. 

Best regard,

### LINKED ISSUE #1535: oculars - compass rose: vertical flip also flips horizontally
### Actual Behaviour
 normal:
image

horizontal flip only (OK):
image

 vertical flip only: NOT OK: (actually this is a 180 rotation):
image

### Steps to reproduce

1. activate oculars
2. select an ocular/telescope combination, eg Plossl 40mm/C11 f/10
3. check that compass rose is active
3. check in the ocular/telescope properties that the combination does not flip the image
2. enable horizontal flip (CTRL+Shift+H) -> observe  that E/W are flipped
3. disable horizontal flip -> observe that N up, S down, E left, W right
3. enable horizontal flip (CTRL+Shift+V)

**Bug:** NS are flipped, but EW are also flipped.

### System
* Stellarium version: Stellarium 0.20.4-817beea [master]
* Operating system: KUbuntu 20.04
* Graphics Card: <Manufacturer (likely Intel, NVidia, AMD?), Model (HD, Geforce, Radeon..., with model number), driver version?>
* Screen type (if applicable): Resolution, HighDPI, scaling

### LINKED ISSUE #1578: Planets and Sun still enlarged when viewed through Oculars
When enlarging the Moon, it still shows the correct size when viewed through the Oculars plugin. However, this is not the case with planets and the sun, and ideally, they would work the same way.

Version 0.21.0.

### LINKED ISSUE #3593: Missing grids and lines after using telrad view
### Actual Behaviour
The grid, constellation and horizon lines disappear after I initiate the telrad view while being in the ocular view.
This only happens if the "Hide grids and lines when enabled" ocular setting is ticked.

Describe or maybe attach a screenshot?

### Steps to reproduce
Start ocular view
While still in ocular view, start telrad view
Stop telrad view
Zoom out and all lines and grids are missing and not responsive.

### System
* Stellarium version: Version 23.4.0
* Operating system: Win 11 Pro
* Graphics Card: Intel UHD Graphics 630. Driver version 31.0.101.2115
* Screen type (if applicable): 1920x1080

### Logfile 
If possible, attach the logfile `log.txt` from your user data directory. Look into the Guide for its location.

### LINKED ISSUE #4126: Oculars Plugin reticle center coordinates do not compensate for "Vertical View Offset"
### Expected Behavior
Center of FOV of Oculars reticle coordinates should compensate for any user change in "Vertical View Offset" in Stellarium's "Sky and Viewing Options".
### Actual Behaviour
Oculars center of FOV does not compensatefor any user change in the "Vertical View Offset", so importing of visually framed targets coordinates into NINA is inaccurate.  degree of inaccuracy varies with the change in overall screen field of view.  The larger the screen FOV, the larger the Ocular center of FOV coordinates inaccuracy.  
Coordinates of Oculars center of FOV reticle are always correct if "Vertical View Offset" in "Sky and Viewing Options" is set to 0% (I believe this is default).
This is a problem when using Stellarium to frame targets and then importing pre-framed coordinates into acquisition software (NINA in my case).
I use a Vertical View Offset to drop horizon down to the bottom of the screen to maximize the sky view on my laptop screen.  My current kludge "fix" is to set the offset at 0% when using Stellarium for framing purposes.  NINA's developers suggested fix is to just use NINA's Framing tab, but Stellarium's GUI is much, much better and preferred.

### Steps to reproduce
Change "Vertical View Offset" to value different than 0%.  Select and center target.  Select Oculars reticle overlay. Zoom screen FOV in or out.  The smaller the screen FOV, the more accurate the Oculars reticle center of FOV becomes with regard to the chosen target; the larger the screen FOV, the less accurate the Oculars reticle center FOV coordinates become with regard to the chosen target. 

### System
* Stellarium version: 24.4 Qt5
* Operating system: Win10 home 22H2, Build 19045.5371
* Graphics Card: Nvidia GeForce
* Screen type (if applicable): Resolution, HighDPI, scaling

### Logfile 
If possible, attach the logfile `log.txt` from your user data directory. Look into the Guide for its location.

### PR DESCRIPTION
**Pull request by gzotti**
### Description

The mode switching and zooming code in the Oculars plugin is hard to follow, with a series of intermingled side effects.
This work here should consolidate the state switching. However, it seems the menu buttons in the top-right GUI panel still need boolean properties. I have moved those as "slave properties" with private getters/setters.

- better understandable mode switching with one centralized method.
- no more entangled side effects
- cleanup not finished yet. For now I have commented away the previous functions for easier comparison. 
- Any missing switching, storage and re-setting of main program settings should be done in the setPluginMode(). 

The other, even larger restructuring is the introduction of a dedicated Finder mode which replaces the hybrid isBinocular() from Ocular mode.  This adds a button to the menu bar. However, Binoculars and Finders are collected in the same class "Finder" given that their properties are specified in the same way. 

Fixes #1509, #356

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

### How Has This Been Tested?

**Test Configuration**:
* Operating system: Windows 10
* Graphics Card: irrelevant

## Checklist:

- [ ] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
> ## 2021: Switching both possible modes off was the original logic. But what if we have an Ocular that is isBinocular()?
> https://github.com/Stellarium/stellarium/blob/b2601a66f250d1b6dcb7f5c0aaebe60af274f7b8/plugins/Oculars/src/Oculars.cpp#L850-L855
> 
> ###### This comment was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in [b2601a6](https://github.com/Stellarium/stellarium/commit/b2601a66f250d1b6dcb7f5c0aaebe60af274f7b8) in #1521. cc @Stellarium.

Binoculars and finderscopes are defined in eyepiece section by the historical reasons. According to  #356 we should introduce new tab in main GUI for finderscopes/binoculars and new button on on-screen panel (+ changes in ini files of course). I tried fix it some time ago, but without success, because fix was required seriously refactoring the code. :( By the fact you are started refactoring the parts of code, which need for the separation of binocular mode also.

**Comment by gzotti** [comment]
Phew, this makes it a bit more ambitious. Then let's introduce a class Finder (maybe derivable from Ocular?) which can be used for binoculars as well (Of course we'll NOT make a funny OO screen mask for binos!) The point is its definition via characterization like "6x30" (mag x aperture) and also TRUE field of view, which may be given in degrees or "x meters in 1km distance" from terrestrial spotting scopes or binos. Do we have to support both? (It's just a arctangent operation. Users can be required to use a pocket calculator to convert their "x meters/km" to degrees, right?)  Actually, the "8x30" is pretty irrelevant to us, only TFOV counts. But in difference to Telrad, we can apply the estimate of limiting magnitudes like for Oculars, here aperture will count again. 

Shall this look like a Telrad plus screen mask plus crosshairs then? Do you have a useful icon? 

If you are in principle OK with this solution, I can remove a lot of comments and TODOs.

**Comment by alex-w** [comment]
> Phew, this makes it a bit more ambitious. Then let's introduce a class Finder (maybe derivable from Ocular?) which can be used for binoculars as well (Of course we'll NOT make a funny OO screen mask for binos!) The point is its definition via characterization like "6x30" (mag x aperture) and also TRUE field of view, which may be given in degrees or "x meters in 1km distance" from terrestrial spotting scopes or binos. Do we have to support both? (It's just a arctangent operation. Users can be required to use a pocket calculator to convert their "x meters/km" to degrees, right?) Actually, the "8x30" is pretty irrelevant to us, only TFOV counts. But in difference to Telrad, we can apply the estimate of limiting magnitudes like for Oculars, here aperture will count again.

Yes, new class Finder is needed here and this class should contain the maths for visualisation the finders/binos

> Shall this look like a Telrad plus screen mask plus crosshairs then? Do you have a useful icon?

Yes and yes (see: https://github.com/Stellarium/stellarium/blob/master/plugins/Oculars/resources/bt_binocular_on.png)

> If you are in principle OK with this solution, I can remove a lot of comments and TODOs.

**Comment by gzotti** [comment]
Apparently superseded by #1633

**Comment by gzotti** [comment]
Don't bother with the conflicts! This requires a rebase first.
